### DISCUSSION TITLE
Questions about how Step and Time works

### DISCUSSION THREAD

**Discussion by maluoi** [StereoKit/StereoKit#614] [Q&A] [answered]
Extracted from #607 for discoverability:

@laultman asked:
> A few questions around the Step method, frames. Time and determinism. The documentation seems to indicate that there really is only ONE step loop and that all UI objects added via AddStepper get called each frame. This would seem to say that a Frame-time is a variable length (non-deterministic) quantity?
>
> SK Time is or seems to be based on system clock time?
> If the above is true, then the more "things" added to the Step cause progressively longer frame time spans thereby reducing the frames per second?
>
> However, in one of my demo apps I have "components" each implemented as IStepper. One of those plays a "flipper" video and uses ssml audio. I have multiple components and they all seem to behave as if each is its own Step.
>
>I am trying to understand the concept and limitations of Step. Why would I use the concept of an Update method if Step is called on each frame for each component that is implemented? Could you explain how the Step is implemented?

**Comment by maluoi** [comment, marked as answer]
## Step Time

The duration of each Step is generally `max( step work time, display refresh speed )`. Ideally, all work within Step is less than the display refresh speed, otherwise you do start running into performance issues! As an example for how long that is, HoloLens 2 is a 60hz display, and one frame is 1/60s, or 16ms.

StereoKit does report this time as `Time.Step` in the preview builds (`Time.Elapsed` in v0.3.6), and it's a critical idea to use this with any time sensitive equations.

## What is Step?

Step and Update are the exact same concept. The words are interchangeable in this context, and you'll sometimes even see me slip up an use the word Update instead 🙄 Unity habits die hard! I chose to say `Step` in StereoKit because it felt far more explicit than `Update`, which is a rather overloaded word.

As to _how_ the Stepping loop is implemented, I actually have the developer control this with their own code! This is the default setup you get with the templates:
```csharp
// Core application loop
while (SK.Step(() =>
{
	if (SK.System.displayType == Display.Opaque)
		Default.MeshCube.Draw(floorMaterial, floorTransform);

	UI.Handle("Cube", ref cubePose, cube.Bounds);
	cube.Draw(cubePose.ToMatrix());
}));
```

Note the use of `while` there! In more recent code I use the `SK.Run` function. I was a little bit sad about hiding the `while` loop, but `SK.Run` allows compatibility with Web/WASM's lifecycle, and is easier to type from scratch. `SK.Run` is still essentially the exact same thing though, a while loop that just keeps calling `SK.Step`. Web is an exception to this, since a while loop like that will hang your browser! So for web, `SK.Run` will just hook up to the browser's animation callback, and Step within that.

Now clearly `SK.Step` is doing additional work here too! But in short: `SK.Step` will loop through all of StereoKit's internal systems, and Step each of those! It's turtles all the way down :)

There's a system for App that's somewhere in the middle of StereoKit's systems, and that App system will invoke the callback provided to Step. This ensures that the dev's code gets called after input has been collected, but before render information has been submitted to the GPU. And if you get there too fast, the rendering system will also pause the whole app until the display is ready to present a new image!

---

Hope that's helpful :)

**Reply by maluoi** [reply]
An interesting bit as well, when StereoKit spits out the performance report to the logs at the end of the application, each line item is a StereoKit internal system! These are sorted by the order in which they Step, so you can kinda get an idea of what else StereoKit is doing in SK.Step, and where the App system is in relation to those things :)

This is running StereoKitTest on my desktop PC in Simulator mode:
```
[SK info] Session Performance Report:
[SK info] ______________________________________________________
[SK info] |         System | Initialize |   Update |  Shutdown |
[SK info] |________________|____________|__________|___________|
[SK info] |       Defaults |    30.92ms |          |    0.05ms |
[SK info] |     FrameBegin |            |  0.074ms |           |
[SK info] |          Input |     0.95ms |  0.053ms |    0.11ms |
[SK info] |        Physics |     0.10ms |  0.029ms |    0.02ms |
[SK info] |       Renderer |     0.37ms |  0.001ms |    0.03ms |
[SK info] |             UI |     6.41ms |  0.013ms |    0.10ms |
[SK info] |       Platform |   381.22ms |          |   19.92ms |
[SK info] |            App |    15.34ms |  1.663ms |           |
[SK info] |      Animation |            |  0.000ms |    0.00ms |
[SK info] |          Lines |     0.08ms |  0.010ms |    0.01ms |
[SK info] |        Sprites |     0.08ms |  0.000ms |    0.02ms |
[SK info] |           Text |            |  0.012ms |    0.01ms |
[SK info] |         UILate |            |  0.000ms |           |
[SK info] |          World |     0.11ms |  0.001ms |    0.00ms |
[SK info] |    FrameRender |            |  0.337ms |           |
[SK info] |         Assets |     0.11ms |  0.002ms |    0.03ms |
[SK info] |          Audio |    63.29ms |  0.005ms |    4.78ms |
[SK info] |________________|____________|__________|___________|
```

**Comment by laultman** [comment]
That is a great answer. That begs another question. How might you implement a parallel process to use the processor cores more effectively (assuming they all return before the current step concludes)? I'm thinking of pipelining some parts of the process.

**Reply by maluoi** [reply]
StereoKit does a little bit of threading itself, primarily for processing Assets! When an image is loaded for example, loading and decoding the file format happen on the asset thread, preventing it from blocking the application. I've also thought it might be interesting to decouple the main Step and the render Step into different threads, this way the app can hang all it wants to, and the visuals will still remain responsive to some degree.

However, you can always still do things like a `Task.Run`, or `async` code if you expect them to take a while! You won't want to do UI or drawing from these threads, but should be able to do plenty of other hard work!

**Comment by laultman** [comment]
I'm currently using the SK.Remoting. I am using a Windows Hosted worker service and a background worker service. All my heavy lifting is running on the background service. My own event management is handling the service communications and the state management. The windows host service is only concerned with the SK. There is no async code in the host service. My application does all the orchestration of objects in the background service. The StateMachine manages the readiness of objects to push into the host service Step method. I am trying to avoid any computations or decision processing inside the loop.
My thought about parallel processing was to use it for drawing. Make one call to Step (the while loop you explained) and inside that single loop step, call all the objects that need to draw (Update) as a parallel foreach instance item. I understand the nature of the hierarchy and the closing of the UI Windows, Handles, etc. because you explained that one earlier.
Your thoughts, please?

**Reply by maluoi** [reply]
Some of this seems fairly reasonable? Some things to note:

You will still need to call `SK.Step` for every screen refresh to get an updated picture, and to drive interactions with the UI. If you don't call `SK.Step`, then StereoKit will do absolutely nothing! (maybe some asset processing) The user's head can move quite a lot in 16ms, and SK does need some way to track hand movement over time for the UI interactions.

I would not recommend using a parallel `for` for drawing. I don't believe SK's `Draw` calls are nicely thread safe! Regardless, calling them from a separate thread could easily cause them to occasionally get double rendered or not rendered at all without careful syncing with the render Step.

16ms is quite a lot of time for a computer, and XR apps are rarely CPU bound! Don't be afraid to do work inside of Step, your optimization efforts are usually best spent on better 3D assets and shaders. It's also important to note that a Step that takes 1ms is no faster than a step that takes 15ms, since the app still needs to wait for the display's refresh.

**Comment by laultman** [comment]
Ahh! So, the screen refresh IS deterministic! It is the standard 15.625ms (~16ms) clock in Windows. I did not connect the screen refresh and the Step relationship. 1 second / 64 = .015625

**Reply by maluoi** [reply]
Screen refresh rate varies depending on the display! HoloLens 2 is 60hz (1/60 = 16.667ms). Quest Pro does either 72hz (1/72 = 13.889ms) or 90hz (1/90 = 11.111ms), and some headsets like the Vive Pro can go up as high as 120hz (1/120 = 8.333ms), might be others that go higher!

There is such a thing as a variable refresh rate, gsync/freesync, but I'm uncertain if that applies to XR yet.

I would not suggest assuming a fixed timestep in SK, even if you are hitting framerate properly! If this is critical for you, you may want to separate your fixed logic from the draw loop to some extent. Physics actually does do this, since a fixed time step increases the stability of physics calculations, but the way this is implemented is with a separate time value, and some logic that will step physics the correct number of times per-frame. It's a little odd, but [you can see how that works here](https://github.com/StereoKit/StereoKit/blob/e72bf7d29db8b1ef79092cdf7dd8e8b851e669b5/StereoKitC/systems/physics.cpp#L95-L98).

**Comment by laultman** [comment]
Thanks Nick, this is really helpful!

**Comment by laultman** [comment]
Got the time stuff working! Cool stuff! Next question, I ended up generating a Model object from scratch to get things to draw in time and space as I wanted. I need to have text attached to elements in the model and in spatial relationship to the model element in question (think a bubble info box). These need to be dynamic. So, I have this eye gaze intersecting with the model bounds. The model is using the eye gaze to track (LookAt) the user. I want the text to be part of the model but I need to change it as the code runs because I am using IoT telemetry data coming from a background worker that is feeding status data to my worker service (SK) process.
There must be a way to do this and position the text in space relative to the person and the object while it tracks the user moving around the physical area (industry shop floor.) Any ideas on this would be greatly appreciated!

**Reply by maluoi** [reply]
Here's an example of a `Model` that always faces the user, with some text anchored to it! I've done the anchoring via matrices here, but you could also do this with the `Hierarchy` class too.

In this particular sample, the GLTF file could contain a node called "TextAnchor", and if that was present, we would just use its transform for the text. If it's not present, it just uses an offset of (0,1,0). If the GLTF has scaling in it though, this could end out a bit weird, but it's not to hard to derive a scale-less transform from the existing node transform.

```csharp
// Probably cache these
Model     model      = Model.FromFile("asset.glb");
ModelNode textAnchor = model?.FindNode("TextAnchor");

// On Step
Vec3   modelPos       = Vec3.Zero;
Matrix modelTransform = Matrix.TR(modelPos, Quat.LookAt(modelPos, Input.Head.position));
Matrix textTransform  = textAnchor?.ModelTransform ?? Matrix.T(0,1,0);

model.Draw(modelTransform);
Text.Add("Info here", textTransform*modelTransform);
```

I hope that answers your question? This probably belongs in a new Discussion post, but I wasn't quite sure how to separate it out.

**Comment by laultman** [comment]
How to separate this out... Well because this directly affects the time/FPS it is worth mentioning this post number. But I agree, I made a definite turn when I learned how the timing worked to a practical implementation discussion, that is mostly in-line code in the step. Maybe we could move this to another discussion thread point back to this one?
